### ISSUE TITLE
Discuss: Background Workers

### ISSUE THREAD

**Issue by peterkeen** [masteringmodernpayments/book#8]
- Learn why processing payments in the background is necessary
- Survey the landscape of background processing systems
- Make charges in the background

**Comment by maxrosecollins** [comment]
Hi

I'm getting an error in the JS

```
undefined local variable or method `permalink' for #<#<Class:0x007febc39c4bc0>:0x007febbd496370>
  (in /Users/max/Sites/dev/sale/app/assets/javascripts/transactions.js.erb)


$.ajax({
        type: "POST",
        url: "/buy/<%= permalink %>",
        data: $('#payment-form').serialize(),


```

**Comment by maxrosecollins** [comment]
any ideas?

**Comment by zenners** [comment]
In order to run sidekiq, do you have to have redis added on to your heroku? 

I get this error in my heroku logs.

Redis::CannotConnectError (Error connecting to Redis on 127.0.0.1:6379 (Errno::ECONNREFUSED)):

I even added a procfile with 

worker: bundle exec sidekiq

**Comment by peterkeen** [comment]
@zenners yeah, you need to have one of the free Redis add ons added to your Heroku app. Then set the `REDIS_PROVIDER` environment variable to the name of the environment variable that your add-on added. For example if you use Redis Cloud you'd add a config to your Heroku like this: `REDIS_PROVIDER=REDISCLOUD_URL`.

**Comment by peterkeen** [comment]
@maxrosecollins as written that needs to be inlined into the page, not in a separate asset file, because it's referring to locals that don't exist when the assets are precompiled.

**Comment by stefanocdn** [comment]
Hi, does someone have an advice on which Redis add-on to choose with Heroku for Sidekiq?

**Comment by peterkeen** [comment]
I personally have had zero problems with either Redis Cloud or Redis To Go.

**Comment by stefanocdn** [comment]
Ok thank you. Is there one with a better pricing / resources?

**Comment by peterkeen** [comment]
It's pretty dependent on your application. For what it's worth, AWS also
offers Redis in the form of ElastiCache, which looks like it runs ~$12/mo
for a 500MB instance. From what I can see it's compatible with Sidekiq.

**Comment by stefanocdn** [comment]
Hi, is it possible to use Puma and Sidekiq on the single Heroku free dyno? Currently my sidekiq jobs are enqueued but not processed, I am using the Procfile and config/puma.rb recommended by Heroku.
Any clue how to make this work for free (for staging test purposes), or do you have necessarily to buy an additionnal heroku worker?

**Comment by dxypher** [comment]
Hey guys,

I cannot get sidekiq to run on Heroku. I have to manually use heroku run bundle exec sidekiq...

I've done the following:
1. heroku addons:create rediscloud
2. heroku config:set REDIS_URL=....
3. added /initializers/redis.rb with:
if ENV["REDISCLOUD_URL"]
  $redis = Redis.new(:url => ENV["REDISCLOUD_URL"])
end
4. add a Procfile with worker: bundle exec sidekiq

Any ideas on getting this working?

Thanks

**Comment by peterkeen** [comment]
Heroku scale worker=1

**Comment by dxypher** [comment]
@peterkeen Cool that worked...can/should this be added to the Procfile or just done manually after deploying?

Thanks

**Comment by batesy** [comment]
For some reason at this point in Sale#create, it is not persisting my sale.amount. It does not persist when I create the sale manually with a block in the console either. Any idea why this might be happening?

**Comment by peco8** [comment]
@peterkeen 
Hi I enjoyed your book so much, which is really helpful.

I have some question about background job.
You said `. The customer will submit the form using AJAX and poll while the background job contacts Stripe and handles the payment details.`

It's good to have background worker to execute some scripts asyncrounouslly.
Here's some question,

If I have some application for SaaS system.
We have to prevent users from using the SaaS system until payment is done.
So in this case we don't really need to have executing those with background job.

What do you think? (I'm not familiar with JS side though.)

**Comment by peco8** [comment]
@dxypher 

Yea I agree that forman is good to go 👍 
If you use docker-compose, it makes your life easier.
